sharedbuffers/mapped_struct.py

```
@@ -812,12 +819,27 @@ def unpack_from(cls, buf, offs, idmap = None):
         buf = _likerobuffer(buf)
         return cls(buf, offs, idmap)
 
-    @cython.locals(obj_offs = cython.ulonglong, dcode = cython.char, index = cython.ulonglong, pindex = "unsigned long *", dataoffs = cython.ulonglong)
-    def __getitem__(self, index):
+
+    @cython.locals(obj_offs = cython.ulonglong, dcode = cython.char, index = cython.longlong, pindex = "unsigned long *",
+        dataoffs = cython.ulonglong, objlen = cython.longlong, xlen = cython.longlong, step = cython.longlong)
+    def _getitem(self, index):
 
         dcode, objlen, itemsize, dataoffs, _struct = self._metadata()
+        xlen = objlen
+        step = 1
+
+        if self.elem_step != 0:
+            step = abs(self.elem_step)
+            if self.elem_step > 0:
+                xlen = (self.elem_end - self.elem_start - 1) / step + 1
+            else:
+                xlen = (self.elem_start - self.elem_end - 1) / step + 1
```

This formula breaks if start==end

---

If start == end, that will be caught by the __getitem__ method, and it will return an empty object.

---

You cannot assume that. People can call the constructor explicitly, you don't expect Python to segfault when you pass a wrong argument, rather raise an error. So you cannot assume input will always be valid if the consecuence of invalid input is a segfault (which is what would happen if xlen is miscomputed)

---

Well, I disagree. If we went that route, we would have to check for buffer integrity everywhere for proxied objects, since they can be constructed that way too.

Regardless, I'll add the check since it's easy to do and not expensive at all.

---

Well, yes. All classes do check for buffer integrity at risky places, AFAIK.